virttest/utils_misc.py

```diff
@@ -3205,12 +3205,94 @@ def verify_dmesg(dmesg_log_file=None, ignore_result=False, level_check=3,
                         4 - emerg,alert,crit,err
                         5 - emerg,alert,crit,err,warn
     :param session: session object to guest
+    :param expected_dmesg: single string comma separated list of
+                           regular expressions whose matches are to be
+                           ignored during verification, e.g. "'x.*', 'y.*'"
     :param return: if ignore_result=True, return True if no errors/crash
                    observed, False otherwise.
     :param raise: if ignore_result=False, raise TestFail exception on
                   observing errors/crash
     """
-    cmd = "dmesg -T -l %s|grep ." % ",".join(map(str, xrange(0, int(level_check))))
+    environ, output, status = __get_kernel_messages(level_check, session)
+    if status == 0:
+        unexpected_messages = __remove_dmesg_matches(output, expected_dmesg)
+        err = __log_full_dmesg(dmesg_log_file, environ, output)
+        if session:
+            session.cmd("dmesg -C")
+        else:
+            process.system("dmesg -C", ignore_status=True)
+        if not ignore_result and unexpected_messages:
+            raise exceptions.TestFail(err)
+        if unexpected_messages:
+            LOG.debug(err)
+        return False
+    return True
+
+
+def __log_full_dmesg(dmesg_log_file, environ, output):
+    """
+    Logs all dmesg messages
+
+    :param dmesg_log_file: if given, messages will be written into this file
+    :param environ: (guest|host)
+    :param output: messages
+    :return: string for test log
+    """
+    err = "Found unexpected failures in %s dmesg log" % environ
+    d_log = "dmesg log:\n%s" % output
+    if dmesg_log_file:
+        with open(dmesg_log_file, "w+") as log_f:
+            log_f.write(d_log)
+        err += " Please check %s dmesg log %s." % (environ, dmesg_log_file)
+    else:
+        err += " Please check %s dmesg log in debug log." % environ
+        LOG.debug(d_log)
+    return err
+
+
+def __remove_dmesg_matches(messages="", expected_dmesg=""):
+    """
+    Removes all messages that match certain regular expressions
+
+    :param messages: single (possibly multi-line) string
+    :param expected_dmesg: single string comma separated list of
+                           regular expressions whose matches are to be
+                           ignored during verification, e.g. "'x.*', 'y.*'"
+    :return: The subset of `messages` that doesn't match
+    """
+    if not expected_dmesg:
+        return messages
+
+    __messages = messages.split('\n')
+    if '' in __messages:
+        __messages.remove('')
+
+    __expected = expected_dmesg.strip("\" ").split(",")
+    expected_messages = [x.strip(" '") for x in __expected]
+    filtered_messages = []
+
+    for expected_regex in expected_messages:
+        not_matching = [x for x in __messages
+                        if not re.findall(expected_regex, x)]
+        filtered_messages = filtered_messages + not_matching
```

Comment:
In my understanding, Here not_matching(list) need satisfy the condition that  it does not match any of expected_messages, we can not just add not_matching into filtered_messages directly after it  just doesn't  match one expected_regex only.

One way to fix this is to add : __messages = filtered_messages after line 3277

Reply:
Thank you you are right, instead I need the intersection.
I fixed this, did some manual testing and also added a unit test case:

Before the fix:
```
>>> messages = "msg_1\nmsg_2\nmsg_3\nmsg_4"
>>> expected_dmesg = "'^msg_1$', '^msg_4$'"
>>> utils_misc.__remove_dmesg_matches(messages, expected_dmesg)
['msg_2', 'msg_3', 'msg_4', 'msg_1', 'msg_2', 'msg_3']
```
After the fix:
```
>>> l1 = [1, 2, 3, 4]
>>> l2 = [ 3, 2 ]
>>> l3 = utils_misc.__intersec(l1, l2)
>>> l3
[2, 3]

>>> messages = "msg_1\nmsg_2\nmsg_3\nmsg_4"
>>> expected_dmesg = "'^msg_1$', '^msg_4$'"
>>> utils_misc.__remove_dmesg_matches(messages, expected_dmesg)
['msg_2', 'msg_3']
```
Unit test run:
```
$ python virttest/unittests/test_utils_misc.py --verbose
...
test__remove_dmesg_matches (__main__.TestDmesgFilter) ... ok

----------------------------------------------------------------------
Ran 1 test in 0.000s

OK
```